vm/src/obj/objset.rs

```diff
@@ -56,536 +55,688 @@ impl PyValue for PyFrozenSet {
     }
 }
 
-pub fn get_elements(obj: &PyObjectRef) -> HashMap<u64, PyObjectRef> {
-    if let Some(set) = obj.payload::<PySet>() {
-        return set.elements.borrow().clone();
-    } else if let Some(frozenset) = obj.payload::<PyFrozenSet>() {
-        return frozenset.elements.clone();
-    }
-    panic!("Not frozenset or set");
+#[derive(Default, Clone)]
+struct PySetInner {
+    elements: HashMap<u64, PyObjectRef>,
 }
 
-fn validate_set_or_frozenset(vm: &VirtualMachine, cls: PyClassRef) -> PyResult<()> {
-    if !(objtype::issubclass(&cls, &vm.ctx.set_type())
-        || objtype::issubclass(&cls, &vm.ctx.frozenset_type()))
-    {
-        return Err(vm.new_type_error(format!("{} is not a subtype of set or frozenset", cls)));
-    }
-    Ok(())
-}
+impl PySetInner {
+    fn new(iterable: OptionalArg<PyObjectRef>, vm: &VirtualMachine) -> PyResult<PySetInner> {
+        let elements: HashMap<u64, PyObjectRef> = match iterable {
+            OptionalArg::Missing => HashMap::new(),
+            OptionalArg::Present(iterable) => {
+                let mut elements = HashMap::new();
+                let iterator = objiter::get_iter(vm, &iterable)?;
+                while let Ok(v) = vm.call_method(&iterator, "__next__", vec![]) {
+                    insert_into_set(vm, &mut elements, &v)?;
+                }
+                elements
+            }
+        };
 
-fn create_set(
-    vm: &VirtualMachine,
-    elements: HashMap<u64, PyObjectRef>,
-    cls: PyClassRef,
-) -> PyResult {
-    if objtype::issubclass(&cls, &vm.ctx.set_type()) {
-        Ok(PyObject::new(
-            PySet {
-                elements: RefCell::new(elements),
-            },
-            PySet::class(vm),
-            None,
-        ))
-    } else if objtype::issubclass(&cls, &vm.ctx.frozenset_type()) {
-        Ok(PyObject::new(
-            PyFrozenSet { elements: elements },
-            PyFrozenSet::class(vm),
-            None,
-        ))
-    } else {
-        Err(vm.new_type_error(format!("{} is not a subtype of set or frozenset", cls)))
+        Ok(PySetInner { elements })
     }
-}
 
-fn perform_action_with_hash(
-    vm: &VirtualMachine,
-    elements: &mut HashMap<u64, PyObjectRef>,
-    item: &PyObjectRef,
-    f: &Fn(&VirtualMachine, &mut HashMap<u64, PyObjectRef>, u64, &PyObjectRef) -> PyResult,
-) -> PyResult {
-    let hash: PyObjectRef = vm.call_method(item, "__hash__", vec![])?;
+    fn len(&self) -> usize {
+        self.elements.len()
+    }
+    fn copy(&self) -> PySetInner {
+        PySetInner {
+            elements: self.elements.clone(),
+        }
+    }
 
-    let hash_value = objint::get_value(&hash);
-    let mut hasher = DefaultHasher::new();
-    hash_value.hash(&mut hasher);
-    let key = hasher.finish();
-    f(vm, elements, key, item)
-}
+    fn contains(&self, needle: PyObjectRef, vm: &VirtualMachine) -> PyResult {
```

Comment:
This should just return a `bool` directly

Reply:
I am not sure how as this can fail.

Reply:
Oh sorry missed the `?` - in this case it should return `PyResult<bool>`

Reply:
Done